### PR TITLE
LPS-121111 make ResourceActionsBag immutable

### PR DESCRIPTION
**Pull request by yuhai**
@dantewang, compare with https://github.com/dantewang/liferay-portal/pull/1767. I add the last three commits to make inner class ResourceActionsBag immutable. Please review this if ci:test run passed.

### CODE REVIEW

**Comment by yuhai** [comment]
ci:test:sf

**Comment by yuhai** [comment]
ci:test:relevant

**Comment by yuhai** [comment]
ci:test

**Comment by dantewang** [comment]
Fetched

**Comment by dantewang** [comment]
ci:test:relevant

**Comment by dantewang** [comment]
@yuhai Did you consider fragment bundles? I haven't look into the changes very carefully, just want to confirm before viewing. Although we want to make the bag immutable, we need to support changes to the bags (by merging the old bag into the new bag and replace the old one with the new one) -- a typical use case would be deploying a fragment bundle I think...

**Comment by yuhai** [comment]
@dantewang, yes, I did. I referred to [LPS-71877](https://issues.liferay.com/browse/LPS-71877) about overriding rule in default.xml.

The rule is as the below:
> To allow for the original use of custom resource action XML files within EXT plugins, we could allow for the second file to take precedence when defining the "*-defaults" and "guest-unsupported", but retain the additive behavior for other elements within the XML file. This would allow us to remove defaults or unsupported permissions with the following steps.

File: portal-impl/src/resource-actions/portal.xml [hunk 1]
```diff
@@ -6,6 +6,7 @@
 		<portlet-name>90</portlet-name>
 		<permissions>
 			<supports>
+				<action-key>ACCESS_IN_CONTROL_PANEL</action-key>
```
Comment (dantewang) [inline portal-impl/src/resource-actions/portal.xml:9]: @yuhai Why is this needed? In LPS-116063 we removed the logic that will add it to 90's action list. I thought it was not needed by 90 given that 90 actually serves as the holder of actions that should actually belong to "root models". `ACCESS_IN_CONTROL_PANEL` should be of a portlet, not a model.

**Comment by dantewang** [comment]
@yuhai Oh, ext file comes first and overrides definitions in the default xml file? This doesn't seem to be the case in master... Is it? Wouldn't master logic read all files and put them together?

File: portal-impl/src/resource-actions/portal.xml (hunk 1 above)
Comment (yuhai) [reply portal-impl/src/resource-actions/portal.xml:9]: @dantewang, got it. I referred to your and tina's branch LPS-112056, LPS-112056 included this. Compared LPS-112056 with the current branch, I thought we lost the behavior and will remove it when moving 90 to other portlet resources. So I added it back. We may remove the commit.

**Comment by yuhai** [comment]
> @yuhai Oh, ext file comes first and overrides definitions in the default xml file? This doesn't seem to be the case in master... Is it? Wouldn't master logic read all files and put them together?

Yes, I think ext comes first by defining the order.
resource.actions.configs=resource-actions/default.xml,resource-actions/default-ext.xml.

It will firstly read default.xml , and then read -ext.xml. 

Check _readResource() method, I think it includes the logic, for "*-defaults" and "guest-unsupported" actions, the logic use clear() method to do(for example,  guestUnsupportedActions.clear()). This will use ext to override to the actions. For the <supports> actions. it will be appended (merge old default.xml and ext.xml together).

**Comment by dantewang** [comment]
@yuhai I see, thanks!

**Comment by dantewang** [comment]
Pull request submitted to https://github.com/dantewang/liferay-portal/pull/1771. [See changes here.](https://github.com/dantewang/liferay-portal/compare/5faed97db664~1...5faed97db664)
